Here I'm Importing some important Libraries

In [1]:
import pandas as pd
import numpy as np

The data we will use is the iris dataset. It is included in scikit_learn in the datasets module. We can load it by calling the load_iris function.

In [2]:
from sklearn.datasets import load_iris
iris_dataset = load_iris()

iris_dataset is similar to a dictionary. Let's check the keys in iris_dataset

In [3]:
print("Keys of iris_dataset:   \n{}".format(iris_dataset.keys()))

Keys of iris_dataset:   
dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename'])


The values of the key DESCR is a short description of the dataset.

In [4]:
print(iris_dataset['DESCR'][:200]+ "\n.....")

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive
.....


The value of the key target_names contains the species of flowers that we want to predict 

In [5]:
print("Target names:   {}".format(iris_dataset['target_names']))

Target names:   ['setosa' 'versicolor' 'virginica']


The value of feature_names contains the features.

In [6]:
print("Feature names:   \n{}".format(iris_dataset['feature_names']))

Feature names:   
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


 The value of data contains numeric measurements of sepal length,sepal width, petal length and petal width .lets check the type of data

In [7]:
print("Type of data:   {}".format(type(iris_dataset['data'])))

Type of data:   <class 'numpy.ndarray'>


The rows of this data array corresponds to the flowers, while the columns represent the four measurements that were taken for each flower.Lets check the shape of data

In [8]:
print("Shape of data:   {}".format(iris_dataset['data'].shape))

Shape of data:   (150, 4)


We see that the array contains measurements for 150 different flowers.Lets check the feature values of the first five samples

In [9]:
print("First five columns of data:\n{}".format(iris_dataset['data'][:5]))

First five columns of data:
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]


the target contains the species of each of the flowers that were measured.Lets check the type of target


In [10]:
print("Type of target:   {}".format(type(iris_dataset['target'])))

Type of target:   <class 'numpy.ndarray'>


Shape of target

In [11]:
print("Shape of target:  {}".format(iris_dataset["target"].shape))

Shape of target:  (150,)


On seeing the target key and exploring the values we can easily see that the species are enclosed as integers from 0 to 2


In [12]:
print("Target: \n{}".format(iris_dataset['target']))

Target: 
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


The meaning of the numbers are given by the iris['target_names']array :0 means setosa, 1 means versicolor and 2 means virginica.

Now we will build a machine learning model.For that we will split the labled data into two parts.One to build our model and other to test the new model.

X represents the features and y represents the labels, random_state rearranges the data

In [13]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split( iris_dataset['data'],
                                iris_dataset['target'],random_state=0)

lets print the shape of X_train,X_test,y_train,y_test

In [14]:
print("X_train shape:  {}".format(X_train.shape))
print("X_test shape:  {}".format(X_test.shape))
print("y_train shape:  {}".format(y_train.shape))
print("y_test shape:  {}".format(y_test.shape))

X_train shape:  (112, 4)
X_test shape:  (38, 4)
y_train shape:  (112,)
y_test shape:  (38,)


k-nearest neighbour machine learning model

In [15]:
from sklearn.neighbors import KNeighborsClassifier
knn= KNeighborsClassifier(n_neighbors=3)

Training the model on the training data

In [16]:
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

lets predict whether oru model is correct

In [17]:
X_new = np.array([[5,2.9,1,0.2]])
print("X_new.shape:  {}".format(X_new.shape))


X_new.shape:  (1, 4)


To make the prediction we call the predict method of the knn object

In [18]:
prediction= knn.predict(X_new)
print("Prediction:  {}".format(prediction))
print("Predicted target name:  {}".format(iris_dataset['target_names'][prediction]))

Prediction:  [0]
Predicted target name:  ['setosa']


Now we predict each iris in the test data

In [19]:
y_pred = knn.predict(X_test)
print("Test set prediction:\n  {}".format(y_pred))

Test set prediction:
  [2 1 0 2 0 2 0 1 1 1 2 1 1 1 1 0 1 1 0 0 2 1 0 0 2 0 0 1 1 0 2 1 0 2 2 1 0
 2]


And we compare it against its labels.And so we get the accuracy of our model

In [20]:
print("Test set score:  {}".format(np.mean(y_pred == y_test)))

Test set score:  0.9736842105263158
